File: Phan_C/bai50/bai_50.jl
using Random
using Printf

function generate_data(n::Int, d::Int, seed::Int)
    Random.seed!(seed)
    return rand(Float64, d, n)
end

function classify(query::Vector{Float64}, train::Matrix{Float64}, d::Int, n::Int)
    min_dist = Inf
    best_idx = -1
    for i in 1:n
        dist = 0.0
        for j in 1:d
            diff = query[j] - train[j, i]
            dist += diff * diff
        end
        if dist < min_dist
            min_dist = dist
            best_idx = i
        end
    end
    return best_idx
end

function main()
    d = 20
    q = 1000
    n_sizes = [1000, 10000, 100000]
    
    queries = generate_data(q, d, 12345)
    
    # JIT warm-up
    train_warmup = generate_data(10, d, 1)
    classify(queries[:, 1], train_warmup, d, 10)
    
    times = Float64[]
    for n in n_sizes
        train = generate_data(n, d, 54321)
        
        t = @elapsed for i in 1:q
            classify(queries[:, i], train, d, n)
        end
        push!(times, t)
    end
    
    ratio1 = times[2] / times[1]
    ratio2 = times[3] / times[2]
    
    @printf("T(10k)/T(1k) = %.2f ; T(100k)/T(10k) = %.2f\n", ratio1, ratio2)
    @printf("n=10^3: %.4fs | n=10^4: %.4fs | n=10^5: %.4fs\n", times[1], times[2], times[3])
end

main()

File: Phan_C/bai50/bai_50.cpp
#include <iostream>
#include <vector>
#include <random>
#include <chrono>
#include <iomanip>
#include <limits>

using namespace std;

vector<vector<double>> generate_data(int n, int d, unsigned int seed) {
    mt19937 rng(seed);
    uniform_real_distribution<double> dist(0.0, 1.0);
    vector<vector<double>> data(n, vector<double>(d));
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < d; ++j) {
            data[i][j] = dist(rng);
        }
    }
    return data;
}

int classify(const vector<double>& query, const vector<vector<double>>& train, int d) {
    double min_dist = numeric_limits<double>::infinity();
    int best_idx = -1;
    for (int i = 0; i < (int)train.size(); ++i) {
        double dist = 0.0;
        for (int j = 0; j < d; ++j) {
            double diff = query[j] - train[i][j];
            dist += diff * diff;
        }
        if (dist < min_dist) {
            min_dist = dist;
            best_idx = i;
        }
    }
    return best_idx;
}

int main() {
    int d = 20;
    int q = 1000;
    vector<int> n_sizes = {1000, 10000, 100000};
    vector<double> times;

    auto queries = generate_data(q, d, 12345);

    volatile int dummy_sink = 0;
    for (int n : n_sizes) {
        auto train = generate_data(n, d, 54321);

        auto start = chrono::high_resolution_clock::now();
        for (const auto& query : queries) {
            dummy_sink += classify(query, train, d);
        }
        auto end = chrono::high_resolution_clock::now();

        chrono::duration<double> elapsed = end - start;
        times.push_back(elapsed.count());
    }

    double ratio1 = times[1] / times[0];
    double ratio2 = times[2] / times[1];

    cout << fixed << setprecision(2);
    cout << "T(10k)/T(1k) = " << ratio1 << " ; T(100k)/T(10k) = " << ratio2 << "\n";
    cout << setprecision(4);
    cout << "n=10^3: " << times[0] << "s | n=10^4: " << times[1] << "s | n=10^5: " << times[2] << "s\n";

    return 0;
}